### PR TITLE
new build all action with tests

### LINKED ISSUE #35: Ledger test not testing the PR version of the containers
When running the ledger test action in a PR, it tests the current master version of the containers but NOT the version from the branch.

### PR DESCRIPTION
**Pull request by mgmgwi**
Created a matrix build to test the images locally before pushing to the GitHub container registry.

This aims to fix #35

### CODE REVIEW

File: .github/workflows/build-all.yml [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+      - name: Cache Docker layers
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ matrix.os_version }}-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-${{ matrix.os_version }}
+            ${{ runner.os }}-buildx
+
+      - name: Lowercase repo owner
```
Comment (udosson) [inline .github/workflows/build-all.yml:50]: `Improvement: Could be replaced by this GHA
https://github.com/ASzc/change-string-case-action

File: .github/workflows/build-all.yml [hunk 2]
```diff
@@ -0,0 +1,241 @@
+name: Matrix Build
+
+on:
+  push:
+    branches: [ 'main' ]
+    tags:
+      - 'v*'
+  pull_request:
+    branches: [ main ]
```
Comment (udosson) [inline .github/workflows/build-all.yml:9]: `Formality`: [ 'main' ]

File: .github/workflows/build-all.yml [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+          echo "::set-output name=n2_unreachable::$UNREACHABLE_N2"
+          echo "::set-output name=n3_unreachable::$UNREACHABLE_N3"
+          echo "::set-output name=n4_unreachable::$UNREACHABLE_N4"
+
+      - name: Safe ledger_state.json for later inspection
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ matrix.os_version }}-tmp-test-${{ github.sha }}-ledger_state.json
+          path: ledger_state.json
+          retention-days: 3
```
Comment (udosson) [inline .github/workflows/build-all.yml:141]: Should we increase this to 5 days?

File: .github/workflows/build-all.yml [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+            ${{ runner.os }}-buildx
+
+      - name: Log in to GitHub Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Lowercase repo owner
```
Comment (udosson) [inline .github/workflows/build-all.yml:199]: `Improvement`: Could be replaced by this GHA
https://github.com/ASzc/change-string-case-action

File: Makefile [hunk 5]
```diff
@@ -1,30 +1,24 @@
...
+ubuntu18:
+	docker build -f "build/Dockerfile.ubuntu18" -t indy-node-container/indy_node:ubuntu18 build
 
 clean_buster:
 	-docker image rm indy-node-container/indy_node:buster
 
-clean_ubuntu_base:
-	-docker image rm indy-node-container/base:ubuntu
+clean_ubuntu16:
+	-docker image rm indy-node-container/indy_node:ubuntu
```
Comment (udosson) [inline Makefile:16]: `ubuntu16`

File: Makefile [hunk 6]
```diff
@@ -1,30 +1,24 @@
...
 
-clean_ubuntu:
+clean_ubuntu18:
 	-docker image rm indy-node-container/indy_node:ubuntu
```
Comment (udosson) [inline Makefile:19]: `ubuntu18`

File: test/.env [hunk 7]
```diff
@@ -1,4 +1,5 @@
 INDY_NETWORK_NAME=idunion_local_test
+IMAGE_NAME=ghcr.io/idunion/indy-node-container/indy_node:buster
```
Comment (udosson) [inline test/.env:2]: Why do we need to reference the `buster` image here?

**Review by udosson — CHANGES_REQUESTED** [review]
@mgmgwi thank you very much for this PR. It really improves the current workflow! Much appreciated! 
I have one major suggestion that would require some changes.
The ideal workflow in my view would be:
`workflow_setup`: Set all ENVs that are being used in the subsequent jobs (e.g., lowercase `repository_owner` only once). All subsequent jobs would depend on this job.
`build`: This job would only cover the building of the three images
`test`: depends on `workflow_setup` and `build`. It contains only the tests (separation of concerns)
`push_to_registry`: depends on `workflow_setup` , `build`, and `test`. No further changes are needed
This could be done in a separate PR.

File: .github/workflows/build-all.yml (hunk 3 above)
Comment (solidnerd) [reply .github/workflows/build-all.yml:141]: I would say here 7 days.

File: .github/workflows/build-all.yml (hunk 1 above)
Comment (solidnerd) [reply .github/workflows/build-all.yml:50]: I wouln't do that that would feeling for me like the lodash approach in CI CD Pipelines.

File: .github/workflows/build-all.yml (hunk 4 above)
Comment (solidnerd) [reply .github/workflows/build-all.yml:199]: I wouln't do that that would feeling for me like the lodash approach in CI CD Pipelines.

**Comment by solidnerd** [comment]
LGTM ! From my side.

**Review by solidnerd — APPROVED** [review]

File: Makefile (hunk 5 above)
Comment (mgmgwi) [reply Makefile]: Done

File: test/.env (hunk 7 above)
Comment (mgmgwi) [reply test/.env:2]: I just wanted to have a sane default.

File: Makefile (hunk 6 above)
Comment (mgmgwi) [reply Makefile]: Done

File: .github/workflows/build-all.yml (hunk 2 above)
Comment (mgmgwi) [reply .github/workflows/build-all.yml]: Done

**Review by udosson — APPROVED** [review]
Thanks for the changes @mgmgwi.
LGTM and ready to merge!

File: .github/workflows/build-all.yml (hunk 4 above)
Comment (mgmgwi) [reply .github/workflows/build-all.yml]: I will leave this for now, the bash command works fine.

**Pull request merged**